### ISSUE TITLE
Highlight notes not getting synced

### ISSUE THREAD

**Issue by righteous-coder** [noembryo/KoHighlights#54]
## Symptom

Some highlights' notes don't get synced.

- ver: 2.3.0.0
- platform: windows 10

## I think this is what happened

1. Create a highlight with no note on device A.
1. Sync highlights between device A and B. The highlight is copied to B. These copies have the same `datetime`.
2. Add add a note on the highlight on device A (or B, doesn't matter). Apparently this does NOT update the `datetime`.
3. Sync highlights again. KoHighlights sees they have the same `datetime` and thus doesn't try to sync the note.

I checked the code and I see it chooses to not sync as long as the highlights have the same `datetime`.

https://github.com/noembryo/KoHighlights/blob/7504ac002a9114059da10f33aac4892123a8090f/main.py#L2583-L2585

## Suggestion

Even if two highlights have the same `datetime`, I want KoHighlights to check the notes and if one of them is empty the non-empty note should be copied. If both have notes then... maybe show a popup window and let the user resolve the conflict.

**Comment by noembryo** [comment]
Well, something is wrong here..
When editing a note, the code of the [`update_comment`](https://github.com/noembryo/KoHighlights/blob/7504ac002a9114059da10f33aac4892123a8090f/main.py#L1531) function updates the annotation's date.

If you don't mean `adding a note` within KOHighlights, but in KOReader, then you might be right.
I haven't check to see if this updates the datetime too..

Adding a lot of `if` in the nested `for` loops that check the annotations, will slow this process a lot, so I want to avoid that as much as possible.
I'll have to think about it a little..

**Comment by righteous-coder** [comment]
> If you don't mean adding a note within KOHighlights, but in KOReader, then you might be right.

Yes, I meant KOReader. I'm sorry, I should've been clear.

**Comment by righteous-coder** [comment]
I feel that you could speed up this function by doing a set intersection operation to find matching annotations instead of doing N^2 comparisons, but I'm not sure

**Comment by noembryo** [comment]
> So I don't think it'd add much overhead if we add the note checking/syncing logic (after the above test).

Yes, you're right.
But I wonder if a better/simpler solution would be to have KOReader change the datetime of the annotation if a note was edited. Perhaps opening an issue about it?
But maybe, I should do it anyway.. We'll see
> I feel that you could speed up this function (`merge_new_highs`) by doing a set intersection operation to find matching annotations instead of doing N^2 comparisons, but I'm not sure

The sets being the `source` and `target` lists?

**Comment by righteous-coder** [comment]
> The sets being the source and target lists?

Yes. The set of source `text`s (str) and the set of target `text`s, specifically.

> Perhaps opening an issue about it?
But maybe, I should do it anyway.. We'll see

That'd be nice. Or I could do it, too.

**Comment by noembryo** [comment]
From the first look at the code, the change you ask is looking very difficult.
> I want KoHighlights to check the notes and if one of them is empty the non-empty note should be copied.

This is easy, but I think is not always ok to do it, since you might want to _delete_ a note.
>If both have notes then... maybe show a popup window and let the user resolve the conflict.

I can't create a popup from this function, because it is used by the CLI too.
> Or I could do it, too.

OK. Perhaps it's better..

**Comment by righteous-coder** [comment]
Created a issue here: https://github.com/koreader/koreader/issues/12957

**Comment by righteous-coder** [comment]
This might be slightly off-topic but can I ask why KoHighlights comapres `text` rather than `pos0` and `pos1`? Is it so that you can sync book files that are not completely the same byte-to-byte but are basically the same book?

**Comment by noembryo** [comment]
No. The books must be the same and there is a md5checksum check to ensure that.
There is a "hack" that let's you sync different ePubs, that the user knows they contain the same book (re-saved by some program and got different md5checksum).

The reason for the `text` check, has more to do with how the original comparison was made in the old format metadata, and maybe also because one check might be better than two..? 😃 
I don't think it's really the faster way though..

**Comment by righteous-coder** [comment]
I see. At least for the new format, maybe it make more sense to check `pos0` and `pos1` because unrelated annotations can have the same `text`? It might be just me but I often highlight a single word so there's a pretty good chance that happens.

**Comment by noembryo** [comment]
Hmm.. This is interesting. I never thought about this test case.
Perhaps it needs to change..

**Comment by noembryo** [comment]
If you can run KOHighlights from the `.py` files (using PySide2), you can check the last commit that might have addressed this issue.
But be careful, haven't really test it yet (not much time)..

**Comment by righteous-coder** [comment]
Thanks a lot. No KOReader release yet since `datetime_updated` got added so I'll build the master and try

**Comment by righteous-coder** [comment]
I struggled a lot with building KOReader but managed to make it work in the end. And yes, your last commit worked. But only notes were updated, not styles or colors. Maybe you could update them too?

**Comment by righteous-coder** [comment]
Also if I send a PR adding test cases for `merge_new_highs`, would you accept it? If yes, can I use pytest, or do you prefer something else?

**Comment by righteous-coder** [comment]
> I don't know where are those are stored

They're stored alongside other fields like this:
```lua
        [1] =     {
            ["chapter"] = "Quickstart guide",
            ["color"] = "yellow",
            ["datetime"] = "2025-01-13 08:29:37",
            ["datetime_updated"] = "2025-01-13 08:29:58",
            ["drawer"] = "lighten",
            ["note"] = "abc",
            ["page"] = "/html/body/h1/text().0",
            ["pageno"] = 1,
            ["pos0"] = "/html/body/h1/text().0",
            ["pos1"] = "/html/body/h1/text().10",
            ["text"] = "Quickstart"
        },
```

**Comment by righteous-coder** [comment]
> Unfortunately I don't do TDD, so I'm not familiar with its practices.

To be clear, I wasn't really suggesting you to employ TDD, rather I thought it'd be nice to add some automatic tests for easy-to-introduce-bugs functions like `merge_new_highs`. You could use GitHub CI to run those tests upon commit but you don't to, you could also just run them manually before pushing.

**Comment by noembryo** [comment]
Check color and style sync in the last commit to `main.py`..

**Comment by righteous-coder** [comment]
Thanks. I just tested and it seems to be working correctly.

**Comment by edo-jan** [comment]
> > If you don't mean adding a note within KOHighlights, but in KOReader, then you might be right.
> 
> Yes, I meant KOReader. I'm sorry, I should've been clear.

I can confirm that this is still an issue. Just added notes to highlights created earlier (in Koreader) and they are not syncing. It would be great if you could fix this as I often return to my highlights and add or change notes made earlier

**Comment by noembryo** [comment]
For this fix to work, you have to run KOHighlights from the source and build the latest KOReader version somehow (@righteous-coder  did just this).
After the KOReader and KOHighlights both get updated, all this will be solved.. 🙂

**Comment by edo-jan** [comment]
> For this fix to work, you have to run KOHighlights from the source and build the latest KOReader version somehow ([@righteous-coder](https://github.com/righteous-coder) did just this). After the KOReader and KOHighlights both get updated, all this will be solved.. 🙂

how different is this from the nightly builds - e.g. on my Kobo I just updated to the latest nightly?

**Comment by noembryo** [comment]
For KOReader, the nightly build should be enough by now (should be installed on all readers that participated in sync).
For KOHighlights though, you have to get the source from GitHub and run this using Python.
This setup should work..

**Comment by edo-jan** [comment]
> For KOReader, the nightly build should be enough by now (must be installed on all readers that participated in sync). For KOHighlights though, you have to get the source from GitHub and run this using Python. This setup should work..

I am afraid this is goes beyond my current competency 🫣... So I should wait for the next version.

**Comment by noembryo** [comment]
If you get the KOReader nightlies somehow to work, I can upload a beta version of KOHighlights somewhere for you..

**Comment by edo-jan** [comment]
> If you get the KOReader nightlies somehow to work, I can upload a beta version of KOHighlights somewhere for you..

Yes, nightlies are easy on Kobo. As to Linux / WSL, folks from the community pointed me to [koreader nightly builds site] (https://gitlab.com/koreader/nightly-builds/-/jobs) .  Kohighlights beta would be great.

**Comment by noembryo** [comment]
If I manage to build one tomorrow (I'm having some strange scripting problems ATM), I'll upload it here..

**Comment by noembryo** [comment]
Try [this beta](https://mega.nz/file/3ZUARbSC#sIvx5Kr10vVSsg6GCcLHge_PkA9eFZvNMoIOtHjcQAE)..

**Comment by edo-jan** [comment]
2.3.0.1 is now synching changes made to highlights on Linux-based Koreader and the archive. However, not synching with the nightly installed on Kobo

**Comment by noembryo** [comment]
Different build perhaps?

**Comment by edo-jan** [comment]
> Different build perhaps?

on linux build 2024.11-203 and on kobo: build 2024.11.197 
Let me try bringing kobo to v203

**Comment by edo-jan** [comment]
Updated all devices to the latest 2024.11-203. Same issue. It syncs between Linux Koreader (this is where the change was made) and the Archive but does not sync with Kobo device (part of the same sync group)

**Comment by noembryo** [comment]
Have to check the metadata, to see what is the difference..

**Comment by edo-jan** [comment]
Let me try making the change on Kobo and see the problem is mirrored - that is if it would sync with the archive but not with Linux

**Comment by edo-jan** [comment]
hmm, it seems that it is working now. I should do some more testing.

**Comment by noembryo** [comment]
Fixed in [v2.3.1.0](https://github.com/noembryo/KoHighlights/releases/tag/v2.3.1.0)
